File: modules/gui/macosx/equalizer.m
/*****************************************************************************
 * equalizer.m: MacOS X interface module
 *****************************************************************************
 * Copyright (C) 2004-2008 the VideoLAN team
 * $Id$
 *
 * Authors: Jérôme Decoodt <djc@videolan.org>
 *          Felix Paul Kühne <fkuehne -at- videolan -dot- org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *****************************************************************************/

/*****************************************************************************
 * Preamble
 *****************************************************************************/
#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <vlc_common.h>
#include <vlc_aout.h>

#include "intf.h"

#include <math.h>

#include "equalizer.h"
#include "../../audio_filter/equalizer_presets.h"

/*****************************************************************************
 * VLCEqualizer implementation
 *****************************************************************************/
@implementation VLCEqualizer

static void ChangeFiltersString( intf_thread_t *p_intf,
                                 char *psz_name, bool b_add )
{
    char *psz_parser, *psz_string;
    int i;
    aout_instance_t *p_aout = getAout();
    vlc_object_t *p_object = VLC_OBJECT(p_aout);
    if( !p_object )
    {
        p_object = vlc_object_hold(pl_Get( p_intf ));
    }

    psz_string = var_GetNonEmptyString( p_object, "audio-filter" );

    if( !psz_string ) psz_string = strdup( "" );

    psz_parser = strstr( psz_string, psz_name );

    if( b_add )
    {
        if( !psz_parser )
        {
            psz_parser = psz_string;
            asprintf( &psz_string, ( *psz_string ) ? "%s,%s" : "%s%s",
                            psz_string, psz_name );
            free( psz_parser );
        }
        else
        {
            vlc_object_release( p_object );
            return;
        }
    }
    else
    {
        if( psz_parser )
        {
            memmove( psz_parser, psz_parser + strlen( psz_name ) +
                            ( *( psz_parser + strlen( psz_name ) ) == ',' ? 1 : 0 ),
                            strlen( psz_parser + strlen( psz_name ) ) + 1 );

            if( *( psz_string+strlen( psz_string ) - 1 ) == ',' )
            {
                *( psz_string+strlen( psz_string ) - 1 ) = '\0';
            }
         }
         else
         {
            free( psz_string );
            vlc_object_release( p_object );
            return;
         }
    }

    aout_EnableFilter( p_object, psz_string, false);
    
    if( (BOOL)config_GetInt( p_object, "macosx-eq-keep" ) == YES )
    {
        /* save changed to config */
        config_PutPsz( p_object, "audio-filter", psz_string );
    }
    
    free( psz_string );
    vlc_object_release( p_object );
}

static bool GetFiltersStatus( intf_thread_t *p_intf,
                                 char *psz_name )
{
    char *psz_parser, *psz_string;
    vlc_object_t *p_object = VLC_OBJECT(getAout());
    if( p_object == NULL )
        p_object = vlc_object_hold(pl_Get( p_intf ));

    if( (BOOL)config_GetInt( p_intf, "macosx-eq-keep" ) == YES )
        psz_string = config_GetPsz( p_intf, "audio-filter" );

    if(! psz_string )
        psz_string = var_GetNonEmptyString( p_object, "audio-filter" );

    vlc_object_release( p_object );

    if( !psz_string ) return false;

    psz_parser = strstr( psz_string, psz_name );

    free( psz_string );

    if ( psz_parser )
        return true;
    else
        return false;
}

- (void)initStrings
{
    int i;
    [o_btn_equalizer setToolTip: _NS("Equalizer")];
    [o_btn_equalizer_embedded setToolTip: _NS("Equalizer")];
    [o_ckb_2pass setTitle: _NS("2 Pass")];
    [o_ckb_2pass setToolTip: _NS("Apply the "
        "equalizer filter twice. The effect will be sharper.")];
    [o_ckb_enable setTitle: _NS("Enable")];
    [o_ckb_enable setToolTip: _NS("Enable the equalizer. Bands can be set "
        "manually or using a preset.")];
    [o_fld_preamp setStringValue: _NS("Preamp")];

    [o_popup_presets removeAllItems];
    for( i = 0; i < 18 ; i++ )
    {
        [o_popup_presets insertItemWithTitle: _NS(preset_list_text[i]) atIndex: i];
    }
    [o_window setTitle: _NS("Equalizer")];

    [self initBandSliders];
}

- (void)equalizerUpdated
{
    intf_thread_t *p_intf = VLCIntf;
    float f_preamp, f_band[10];
    char *psz_bands, *psz_bands_init, *p_next;
    bool b_2p;
    int i;
    bool b_enabled = GetFiltersStatus( p_intf, (char *)"equalizer" );
    vlc_object_t *p_object = VLC_OBJECT(getAout());

    if( p_object == NULL )
        p_object = vlc_object_hold(pl_Get( p_intf ));

    var_Create( p_object, "equalizer-preamp", VLC_VAR_FLOAT |
                VLC_VAR_DOINHERIT );
    var_Create( p_object, "equalizer-bands", VLC_VAR_STRING |
                VLC_VAR_DOINHERIT );

    psz_bands = var_GetNonEmptyString( p_object, "equalizer-bands" );

    if( psz_bands == NULL )
        psz_bands = strdup( "0 0 0 0 0 0 0 0 0 0" );

    if( (BOOL)config_GetInt( p_intf, "macosx-eq-keep" ) == YES )
    {
        b_2p = (BOOL)config_GetInt( p_object, "equalizer-2pass" );
        f_preamp = config_GetFloat( p_object, "equalizer-preamp" );
    }
    else
    {
        b_2p = var_GetBool( p_object, "equalizer-2pass" );
        f_preamp = var_GetFloat( p_object, "equalizer-preamp" );
    }

    vlc_object_release( p_object );
    
    /* Set the preamp slider */
    [o_slider_preamp setFloatValue: f_preamp];

    /* Set the bands slider */
    psz_bands_init = psz_bands;

    for( i = 0; i < 10; i++ )
    {
        /* Read dB -20/20 */
        f_band[i] = strtof( psz_bands, &p_next );
        if( !p_next || p_next == psz_bands ) break; /* strtof() failed */
    
        if( !*psz_bands ) break; /* end of line */
        psz_bands = p_next+1;
    }
    free( psz_bands_init );
    [self setBandSlidersValues:f_band];

    /* Set the the checkboxes */
    [o_ckb_enable setState: b_enabled];

    [o_ckb_2pass setState: b_2p];        
}

- (IBAction)bandSliderUpdated:(id)sender
{
    intf_thread_t *p_intf = VLCIntf;
    vlc_object_t *p_object = VLC_OBJECT(getAout());

    if( p_object == NULL )
        p_object = vlc_object_hold(pl_Get( p_intf ));

    const char *psz_values;
    NSString *preset = [NSString stringWithFormat:@"%.1f", [o_slider_band1 floatValue] ];
    preset = [preset stringByAppendingFormat:@"%.1f ", [o_slider_band2 floatValue] ];
    preset = [preset stringByAppendingFormat:@"%.1f ", [o_slider_band3 floatValue] ];
    preset = [preset stringByAppendingFormat:@"%.1f ", [o_slider_band4 floatValue] ];
    preset = [preset stringByAppendingFormat:@"%.1f ", [o_slider_band5 floatValue] ];
    preset = [preset stringByAppendingFormat:@"%.1f ", [o_slider_band6 floatValue] ];
    preset = [preset stringByAppendingFormat:@"%.1f ", [o_slider_band7 floatValue] ];
    preset = [preset stringByAppendingFormat:@"%.1f ", [o_slider_band8 floatValue] ];
    preset = [preset stringByAppendingFormat:@"%.1f ", [o_slider_band9 floatValue] ];
    preset = [preset stringByAppendingFormat:@"%.1f", [o_slider_band10 floatValue] ];

    psz_values = [preset UTF8String];
    var_SetString( p_object, "equalizer-bands", psz_values );

    if( (BOOL)config_GetInt( p_intf, "macosx-eq-keep" ) == YES )
    {
        /* save changed to config */
        config_PutPsz( p_intf, "equalizer-bands", psz_values );

        /* save to vlcrc */
        config_SaveConfigFile( p_intf, "equalizer" );
    }

    vlc_object_release( p_object );
}

- (IBAction)changePreset:(id)sender
{
    intf_thread_t *p_intf = VLCIntf;
    int i;
    vlc_object_t *p_object= VLC_OBJECT(getAout());
    if( p_object == NULL )
        p_object = vlc_object_hold(pl_Get( p_intf ));

    var_SetString( p_object , "equalizer-preset" , preset_list[[sender indexOfSelectedItem]] );

    NSString *preset = @"";
    const char *psz_values;
    for( i = 0; i < 10; i++ )
    {
        preset = [preset stringByAppendingFormat:@"%.1f ", eqz_preset_10b[[sender indexOfSelectedItem]]->f_amp[i] ];
    }
    psz_values = [preset UTF8String];
    var_SetString( p_object, "equalizer-bands", psz_values );
    var_SetFloat( p_object, "equalizer-preamp", eqz_preset_10b[[sender indexOfSelectedItem]]->f_preamp);

    [o_slider_preamp setFloatValue: eqz_preset_10b[[sender indexOfSelectedItem]]->f_preamp];

    [self setBandSlidersValues:(float *)eqz_preset_10b[[sender indexOfSelectedItem]]->f_amp];

    if( (BOOL)config_GetInt( p_intf, "macosx-eq-keep" ) == YES )
    {
        /* save changed to config */
        config_PutPsz( p_intf, "equalizer-bands", psz_values );
        config_PutFloat( p_intf, "equalizer-preamp", eqz_preset_10b[[sender indexOfSelectedItem]]->f_preamp );
        config_PutPsz( p_intf, "equalizer-preset", preset_list[[sender indexOfSelectedItem]] );

        /* save to vlcrc */
        config_SaveConfigFile( p_intf, "equalizer" );
    }

    vlc_object_release( p_object );
}

- (IBAction)enable:(id)sender
{
    ChangeFiltersString( VLCIntf, (char *)"equalizer", [sender state] );
}

- (IBAction)preampSliderUpdated:(id)sender
{
    intf_thread_t *p_intf = VLCIntf;
    float f_preamp = [sender floatValue] ;

    vlc_object_t *p_object = VLC_OBJECT(getAout());
    if( p_object == NULL )
        p_object = vlc_object_hold(pl_Get( p_intf ));

    var_SetFloat( p_object, "equalizer-preamp", f_preamp );

    if( (BOOL)config_GetInt( p_intf, "macosx-eq-keep" ) == YES )
    {
        /* save changed to config */
        config_PutFloat( p_intf, "equalizer-preamp", f_preamp );

        /* save to vlcrc */
        config_SaveConfigFile( p_intf, "equalizer" );
    }

    vlc_object_release( p_object );
}

- (IBAction)toggleWindow:(id)sender
{
    if( [o_window isVisible] )
    {
        [o_window orderOut:sender];
        [o_btn_equalizer setState:NSOffState];
        [o_btn_equalizer_embedded setState:NSOffState];
    }
    else
    {
        [o_window makeKeyAndOrderFront:sender];
        [o_btn_equalizer setState:NSOnState];
        [o_btn_equalizer_embedded setState:NSOnState];
    }
}

- (IBAction)twopass:(id)sender
{
    intf_thread_t *p_intf = VLCIntf;
    bool b_2p = [sender state] ? true : false;
    aout_instance_t *p_aout = getAout();
    vlc_object_t *p_object= VLC_OBJECT(p_aout);
    if( p_object == NULL )
        p_object = vl_object_hold(pl_Get( p_intf ));

    var_SetBool( p_object, "equalizer-2pass", b_2p );

    if( (BOOL)config_GetInt( p_intf, "macosx-eq-keep" ) == YES )
    {
        /* save changed to config */
        config_PutInt( p_intf, "equalizer-2pass", (int)b_2p );

        /* save to vlcrc */
        config_SaveConfigFile( p_intf, "equalizer" );
    }

    vlc_object_release( p_object );
}

- (void)windowWillClose:(NSNotification *)aNotification
{
    [o_btn_equalizer setState: NSOffState];
}

- (void)awakeFromNib
{
    int i;
    vlc_object_t *p_object= VLC_OBJECT(getAout());
    if( p_object == NULL )
        p_object = vlc_object_hold(pl_Get( VLCIntf ));

    [o_window setExcludedFromWindowsMenu: TRUE];

    [self initStrings];

    if( p_object )
    {
        char *psz_preset;

        var_Create( p_object, "equalizer-preset", VLC_VAR_STRING |
                    VLC_VAR_DOINHERIT );
        psz_preset = var_GetNonEmptyString( p_object, "equalizer-preset" );

        for( i = 0 ; (psz_preset != NULL) && (i < 18) ; i++ )
        {
            if( strcmp( preset_list[i], psz_preset ) )
                continue;
    
            [o_popup_presets selectItemAtIndex: i];
        

            [o_slider_preamp setFloatValue: eqz_preset_10b[i]->f_preamp];
            [self setBandSlidersValues: (float *)eqz_preset_10b[i]->f_amp];

            if( strcmp( psz_preset, "flat" ) )
            {
                char psz_bands[100];
    
                snprintf( psz_bands, sizeof( psz_bands ),
                          "%.1f %.1f %.1f %.1f %.1f %.1f %.1f "
                          "%.1f %.1f %.1f",
                          eqz_preset_10b[i]->f_amp[0],
                          eqz_preset_10b[i]->f_amp[1],
                          eqz_preset_10b[i]->f_amp[2],
                          eqz_preset_10b[i]->f_amp[3],
                          eqz_preset_10b[i]->f_amp[4],
                          eqz_preset_10b[i]->f_amp[5],
                          eqz_preset_10b[i]->f_amp[6],
                          eqz_preset_10b[i]->f_amp[7],
                          eqz_preset_10b[i]->f_amp[8],
                          eqz_preset_10b[i]->f_amp[9] );
    
                var_Create( p_object, "equalizer-preamp", VLC_VAR_FLOAT |
                            VLC_VAR_DOINHERIT );
                var_Create( p_object, "equalizer-bands", VLC_VAR_STRING |
                            VLC_VAR_DOINHERIT );
                var_SetFloat( p_object, "equalizer-preamp",
                              eqz_preset_10b[i]->f_preamp );
                var_SetString( p_object, "equalizer-bands", psz_bands );
            }
        }
        free( psz_preset );
        vlc_object_release( p_object );
    }

    [self equalizerUpdated];    
}


- (id)sliderByIndex:(int)index
{
    switch(index)
    {
        case 0 : return o_slider_band1;
        case 1 : return o_slider_band2;
        case 2 : return o_slider_band3;
        case 3 : return o_slider_band4;
        case 4 : return o_slider_band5;
        case 5 : return o_slider_band6;
        case 6 : return o_slider_band7;
        case 7 : return o_slider_band8;
        case 8 : return o_slider_band9;
        case 9 : return o_slider_band10;
        default : return nil;
    }
}

- (void)setBandSlidersValues:(float *)values
{
    int i = 0;
    for (i = 0 ; i<= 9 ; i++)
    {
        [self setValue:values[i] forSlider:i];
    }
}

- (void)initBandSliders
{
    int i = 0;
    for (i = 0 ; i< 9 ; i++)
    {
        [self setValue:0.0 forSlider:i];
    }
}

- (void)setValue:(float)value forSlider:(int)index
{
    id slider = [self sliderByIndex:index];

    if (slider != nil)
    {
        [slider setFloatValue:value];
    }
}

@end

File: modules/gui/qt4/components/equalizer.cpp
/*****************************************************************************
 * equalizer.cpp : Equalizer
 ****************************************************************************
 * Copyright (C) 2006 the VideoLAN team
 * $Id: preferences.cpp 16643 2006-09-13 12:45:46Z zorglub $
 *
 * Authors: Clément Stenac <zorglub@videolan.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *****************************************************************************/

#include <QLabel>
#include <QVariant>
#include <QString>
#include <QFont>
#include <QGridLayout>

#include "components/equalizer.hpp"
#include "qt4.hpp"

#include "../../audio_filter/equalizer_presets.h"
#include <audio_output.h>
#include <aout_internal.h>
#include <vlc_intf_strings.h>
#include <assert.h>

static const QString band_frequencies[] =
{
    "   60Hz  ", " 170 Hz " , " 310 Hz ", " 600 Hz ", "  1 kHz  ",
    "  3 kHz  " , "  6 kHz ", " 12 kHz ", " 14 kHz ", " 16 kHz "
};

Equalizer::Equalizer( intf_thread_t *_p_intf, QWidget *_parent ) :
                            QWidget( _parent ) , p_intf( _p_intf )
{
    QFont smallFont = QApplication::font(0);
    smallFont.setPointSize( smallFont.pointSize() - 3 );

    ui.setupUi( this );

    ui.preampLabel->setFont( smallFont );
    ui.preampSlider->setMaximum( 400 );
    for( int i = 0 ; i < NB_PRESETS ; i ++ )
    {
        ui.presetsCombo->addItem( qfu( preset_list_text[i] ),
                                  QVariant( i ) );
    }
    CONNECT( ui.presetsCombo, activated( int ), this, setPreset( int ) );

    BUTTONACT( ui.enableCheck, enable() );
    BUTTONACT( ui.eq2PassCheck, set2Pass() );

    CONNECT( ui.preampSlider, valueChanged(int), this, setPreamp() );

    QGridLayout *grid = new QGridLayout( ui.frame );
    grid->setMargin( 0 );
    for( int i = 0 ; i < BANDS ; i++ )
    {
        bands[i] = new QSlider( Qt::Vertical );
        bands[i]->setMaximum( 400 );
        CONNECT( bands[i], valueChanged(int), this, setBand() );
        band_texts[i] = new QLabel( band_frequencies[i] + "\n0.0dB" );
        band_texts[i]->setFont( smallFont );
        grid->addWidget( bands[i], 0, i );
        grid->addWidget( band_texts[i], 1, i );
    }

    /* Write down initial values */
    aout_instance_t *p_aout = (aout_instance_t *)vlc_object_find(p_intf,
                                    VLC_OBJECT_AOUT, FIND_ANYWHERE);
    char *psz_af = NULL;
    char *psz_bands;
    float f_preamp;
    if( p_aout )
    {
        psz_af = var_GetString( p_aout, "audio-filter" );
        if( var_GetBool( p_aout, "equalizer-2pass" ) )
            ui.eq2PassCheck->setChecked( true );
        psz_bands = var_GetString( p_aout, "equalizer-bands" );
        f_preamp = var_GetFloat( p_aout, "equalizer-preamp" );
        vlc_object_release( p_aout );
    }
    else
    {
        psz_af = config_GetPsz( p_intf, "audio-filter" );
        if( config_GetInt( p_intf, "equalizer-2pass" ) )
            ui.eq2PassCheck->setChecked( true );
        psz_bands = config_GetPsz( p_intf, "equalizer-bands" );
        f_preamp = config_GetFloat( p_intf, "equalizer-preamp" );
    }
    if( psz_af && strstr( psz_af, "equalizer" ) != NULL )
        ui.enableCheck->setChecked( true );
    enable( ui.enableCheck->isChecked() );

    setValues( psz_bands, f_preamp );
}

Equalizer::~Equalizer()
{
}

void Equalizer::enable()
{
    bool en = ui.enableCheck->isChecked();
    aout_EnableFilter( VLC_OBJECT( p_intf ), "equalizer",
                       en ? VLC_TRUE : VLC_FALSE );
    enable( en );
}

void Equalizer::enable( bool en )
{
    ui.eq2PassCheck->setEnabled( en );
    ui.preampLabel->setEnabled( en );
    ui.preampSlider->setEnabled( en  );
    for( int i = 0 ; i< BANDS; i++ )
    {
        bands[i]->setEnabled( en ); band_texts[i]->setEnabled( en );
    }
}

void Equalizer::set2Pass()
{
    aout_instance_t *p_aout= (aout_instance_t *)vlc_object_find(p_intf,
                                 VLC_OBJECT_AOUT, FIND_ANYWHERE);
    vlc_bool_t b_2p = ui.eq2PassCheck->isChecked();

    if( p_aout == NULL )
        config_PutInt( p_intf, "equalizer-2pass", b_2p );
    else
    {
        var_SetBool( p_aout, "equalizer-2pass", b_2p );
        config_PutInt( p_intf, "equalizer-2pass", b_2p );
        for( int i = 0; i < p_aout->i_nb_inputs; i++ )
        {
            p_aout->pp_inputs[i]->b_restart = VLC_TRUE;
        }
        vlc_object_release( p_aout );
    }
}

void Equalizer::setPreamp()
{
    float f= (float)(  ui.preampSlider->value() ) /10 - 20;
    char psz_val[5];
    aout_instance_t *p_aout= (aout_instance_t *)vlc_object_find(p_intf,
                                       VLC_OBJECT_AOUT, FIND_ANYWHERE);

    sprintf( psz_val, "%.1f", f );
    ui.preampLabel->setText( qtr("Preamp\n") + psz_val + qtr("dB") );
    if( p_aout )
    {
        delCallbacks( p_aout );
        var_SetFloat( p_aout, "equalizer-preamp", f );
        addCallbacks( p_aout );
        vlc_object_release( p_aout );
    }
    config_PutFloat( p_intf, "equalizer-preamp", f );
}

void Equalizer::setBand()
{
    char psz_values[102]; memset( psz_values, 0, 102 );

    /**\todo smoothing */

    for( int i = 0 ; i< BANDS ; i++ )
    {
        char psz_val[5];
        float f_val = (float)(  bands[i]->value() ) / 10 - 20 ;
        sprintf( psz_values, "%s %f", psz_values, f_val );
        sprintf( psz_val, "% 5.1f", f_val );
        band_texts[i]->setText( band_frequencies[i] + "\n" + psz_val + "dB" );
    }
    aout_instance_t *p_aout= (aout_instance_t *)vlc_object_find(p_intf,
                                          VLC_OBJECT_AOUT, FIND_ANYWHERE);
    if( p_aout )
    {
        delCallbacks( p_aout );
        var_SetString( p_aout, "equalizer-bands", psz_values );
        addCallbacks( p_aout );
        vlc_object_release( p_aout );
    }
}
void Equalizer::setValues( char *psz_bands, float f_preamp )
{
    char *p = psz_bands;
    for( int i = 0; i < 10; i++ )
    {
        char psz_val[5];

        float f = strtof( p, &p );
        int  i_val= (int)( ( f + 20 ) * 10 );
        bands[i]->setValue(  i_val );

        sprintf( psz_val, "% 5.1f", f );
        band_texts[i]->setText( band_frequencies[i] + "\n" + psz_val + "dB" );

        if( p == NULL ) break;
        p++;
        if( *p == 0 )  break;
    }
    char psz_val[5];
    int i_val = (int)( ( f_preamp + 20 ) * 10 );
    sprintf( psz_val, "%.1f", f_preamp );
    ui.preampSlider->setValue( i_val );
    ui.preampLabel->setText( qtr("Preamp\n") + psz_val + qtr("dB") );
}

void Equalizer::setPreset( int preset )
{
    aout_instance_t *p_aout= (aout_instance_t *)vlc_object_find(p_intf,
                                                VLC_OBJECT_AOUT, FIND_ANYWHERE);

    char psz_values[102]; memset( psz_values, 0, 102 );
    for( int i = 0 ; i< 10 ;i++ )
        sprintf( psz_values, "%s %.1f", psz_values,
                                        eqz_preset_10b[preset]->f_amp[i] );

    if( p_aout )
    {
        delCallbacks( p_aout );
        var_SetString( p_aout, "equalizer-bands", psz_values );
        var_SetFloat( p_aout, "equalizer-preamp",
                      eqz_preset_10b[preset]->f_preamp );
        addCallbacks( p_aout );
        vlc_object_release( p_aout );
    }
    config_PutPsz( p_intf, "equalizer-bands", psz_values );
    config_PutFloat( p_intf, "equalizer-preamp",
                    eqz_preset_10b[preset]->f_preamp );

    setValues( psz_values, eqz_preset_10b[preset]->f_preamp );
}

void Equalizer::delCallbacks( aout_instance_t *p_aout )
{
//    var_DelCallback( p_aout, "equalizer-bands", EqzCallback, this );
//    var_DelCallback( p_aout, "equalizer-preamp", EqzCallback, this );
}

void Equalizer::addCallbacks( aout_instance_t *p_aout )
{
//    var_AddCallback( p_aout, "equalizer-bands", EqzCallback, this );
//    var_AddCallback( p_aout, "equalizer-preamp", EqzCallback, this );
}
